kube-runtime/src/reflector/object_ref.rs

```diff
@@ -104,102 +100,68 @@ impl<K: Resource> ObjectRef<K> {
-    /// Convert into a reference to `K2`
-    ///
-    /// Note that no checking is done on whether this conversion makes sense. For example, every `Service`
-    /// has a corresponding `Endpoints`, but it wouldn't make sense to convert a `Pod` into a `Deployment`.
     #[must_use]
-    pub fn into_kind_unchecked<K2: Resource>(self, dt2: K2::DynamicType) -> ObjectRef<K2> {
-        ObjectRef {
-            dyntype: dt2,
-            name: self.name,
-            namespace: self.namespace,
-            extra: self.extra,
-        }
-    }
-
-    pub fn erase(self) -> ObjectRef<DynamicObject> {
-        ObjectRef {
-            dyntype: kube_client::api::ApiResource::erase::<K>(&self.dyntype),
-            name: self.name,
-            namespace: self.namespace,
-            extra: self.extra,
-        }
+    pub fn with_types(mut self, types: &TypeMeta) -> Self {
+        self.types = Some(types.clone());
+        self
     }
 }
 
-impl<K: Resource> From<ObjectRef<K>> for ObjectReference {
-    fn from(val: ObjectRef<K>) -> Self {
-        let ObjectRef {
-            dyntype: dt,
-            name,
-            namespace,
-            extra: Extra {
+#[derive(Debug, Error)]
+#[error("failed to find type information")]
+/// Failed to parse group version
+pub struct MissingTypeInfo;
+
+
+impl TryFrom<ObjectRef> for ObjectReference {
+    type Error = MissingTypeInfo;
+
+    fn try_from(val: ObjectRef) -> Result<Self, Self::Error> {
+        if let Some(t) = &val.types {
+            let ObjectRef {
+                name,
+                namespace,
+                extra:
+                    Extra {
+                        resource_version,
+                        uid,
+                    },
+                ..
+            } = val;
+            Ok(ObjectReference {
+                api_version: Some(t.api_version.clone()),
+                kind: Some(t.kind.clone()),
+                field_path: None,
+                name: Some(name),
+                namespace,
                 resource_version,
                 uid,
-            },
-        } = val;
-        ObjectReference {
-            api_version: Some(K::api_version(&dt).into_owned()),
-            kind: Some(K::kind(&dt).into_owned()),
-            field_path: None,
-            name: Some(name),
-            namespace,
-            resource_version,
-            uid,
+            })
+        } else {
+            Err(MissingTypeInfo)
         }
     }
 }
 
-impl<K: Resource> Display for ObjectRef<K> {
+impl Display for ObjectRef {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(
-            f,
-            "{}.{}.{}/{}",
-            K::kind(&self.dyntype),
-            K::version(&self.dyntype),
-            K::group(&self.dyntype),
-            self.name
-        )?;
+        if let Some(tm) = &self.types {
+            write!(f, "{}.{}/{}", tm.kind, tm.api_version, self.name)?;
+        } else {
+            write!(f, "{}", self.name)?;
```

Comment:
I think I'd rather at least have some kind of `unknown/{}` here to indicate that something went wrong, and that the type info is missing.

But ideally this should just not be possible in the first place.

Reply:
Have added an `unknown` prefix.

I agree the situation isn't ideal, but there should not be any sensible real world scenarios where it is missing. Sure, you can construct an objectref without types, but it's also not the most user-friendly type to send out to the world. Maybe we can find a way to keep it internal since it is so tied to controller-logic?